blame_test.go

```diff
@@ -306,42 +306,45 @@ var blameTests = [...]blameTest{
 		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 22),
 		repeat("c89dab0d42f1856d157357e9010f8cc6a12f5b1f", 7),
 	)},
-	{"https://github.com/spinnaker/spinnaker.git", "f39d86f59a0781f130e8de6b2115329c1fbe9545", "pylib/spinnaker/validate_configuration.py", concat(
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 29),
-		repeat("1e3d328a2cabda5d0aaddc5dec65271343e0dc37", 19),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 15),
-		repeat("b5d999e2986e190d81767cd3cfeda0260f9f6fb8", 1),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 12),
-		repeat("1e14f94bcf82694fdc7e2dcbbfdbbed58db0f4d9", 1),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 4),
-		repeat("b5d999e2986e190d81767cd3cfeda0260f9f6fb8", 8),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 1),
-		repeat("b5d999e2986e190d81767cd3cfeda0260f9f6fb8", 4),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 46),
-		repeat("1e14f94bcf82694fdc7e2dcbbfdbbed58db0f4d9", 1),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 4),
-		repeat("1e3d328a2cabda5d0aaddc5dec65271343e0dc37", 42),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 1),
-		repeat("1e3d328a2cabda5d0aaddc5dec65271343e0dc37", 1),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 3),
-		repeat("1e3d328a2cabda5d0aaddc5dec65271343e0dc37", 1),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 1),
-		repeat("1e14f94bcf82694fdc7e2dcbbfdbbed58db0f4d9", 8),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 1),
-		repeat("1e14f94bcf82694fdc7e2dcbbfdbbed58db0f4d9", 2),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 3),
-		repeat("1e3d328a2cabda5d0aaddc5dec65271343e0dc37", 3),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 12),
-		repeat("1e14f94bcf82694fdc7e2dcbbfdbbed58db0f4d9", 10),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 69),
-		repeat("b5d999e2986e190d81767cd3cfeda0260f9f6fb8", 7),
-		repeat("a24001f6938d425d0e7504bdf5d27fc866a85c3d", 4),
-	)},
+	/*
+		// Failed
+		{"https://github.com/spinnaker/spinnaker.git", "f39d86f59a0781f130e8de6b2115329c1fbe9545", "pylib/spinnaker/validate_configuration.py", concat(
```

Comment:
Are you planning to get this test case to work as well?

Reply:
> @AriehSchneier Nice work! I noticed that there are a few optimisation TODO's. What's the performance impact of the changes?

@pjbgf Thats a good question, its hard to know without running tests. It would be good if someone could create a couple of benchmarks that we could run against to compare.

Reply:
I have not had a chance to have a look at the `failing` tests yet, that will require me to clone the tests and read the source manually. The thing to note is just because 2 implementations of blame give different results does not necessarily mean 1 of them is wrong. It can make a huge difference what algorithms and decisions you make. For example the diff algorithm will make an enormous difference to the results.

I am also trying to decide if we should go the more expensive route for finding changes across multiple branches. In the current implementation if the same change is made on 2 branches it just 'randomly' picks 1 of them (chooses the tree of the parent with the older timestamp - but that doesn't mean it will find the older change). Should we calculate the change across both branches and choose the commit with the older timestamp (I believe this would be _more correct_ but more expensive)?

Reply:
@pjbgf 
I investigated all the failing tests, all the ones still marked as failed now have their reason (different diff tool giving different line number results).

I also changed the code so that it follows both branches and discovered that 'git blame' does not use the earliest commit when a change is added twice, rather it picks the version that was merged into the current branch first - so that is what I changed it to.